crates/compiler/intern/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+//! Generic interners for concurrent and single-thread use cases.
+
+use std::{
+    cell::RefCell,
+    hash::{BuildHasher, Hash, Hasher},
+    sync::Arc,
+};
+
+use parking_lot::{Mutex, RwLock};
+use roc_collections::{default_hasher, BumpMap};
+
+/// An interned value.
+///
+/// When possible, prefer comparing/hashing on the [Interned] representation of a value, rather
+/// than the value itself.
+#[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Interned<T>(usize, std::marker::PhantomData<T>);
+impl<T> Clone for Interned<T> {
+    fn clone(&self) -> Self {
+        Self(self.0, Default::default())
+    }
+}
+
+impl<T> Copy for Interned<T> {}
+
+/// A concurrent interner, suitable for usage between threads.
+///
+/// The interner does not currently maintain its own arena; you will have to supply
+/// values-to-be-interned as allocated in an independent arena.
+///
+/// If you need a concurrent global interner, you'll likely want each thread to take a
+/// [ThreadLocalInterner] via [GlobalInterner::fork], for caching purposes.
+///
+/// Originally derived from https://gist.github.com/matklad/44ba1a5a6168bc0c26c995131c007907;
+/// thank you, Aleksey!
+#[derive(Debug)]
+pub struct GlobalInterner<'a, K> {
+    map: Mutex<BumpMap<&'a K, Interned<K>>>,
+    vec: RwLock<Vec<&'a K>>,
+}
+
+/// A derivative of a [GlobalInterner] interner that provides caching desirable for
+/// thread-local workloads. The only way to get a [ThreadLocalInterner] is via
+/// [GlobalInterner::fork].
+///
+/// All values interned into a [ThreadLocalInterner] are made available in its parent
+/// [GlobalInterner], making this suitable for global sharing of interned values.
+///
+/// Originally derived from https://gist.github.com/matklad/44ba1a5a6168bc0c26c995131c007907;
+/// thank you, Aleksey!
+#[derive(Debug)]
+pub struct ThreadLocalInterner<'a, K> {
+    parent: Arc<GlobalInterner<'a, K>>,
+    map: BumpMap<&'a K, Interned<K>>,
```

Comment:
we should investigate `FnvHashmap` here too

Reply:
I think Fnv won't help too much, since we're hashing layouts here (and not ints)